DOCS-938: Document Vision Service API (#1567)
viamrobotics/docs: docs/services/vision/_index.md

[`GetObjectPointClouds`](#getobjectpointclouds) | Get a list of point cloud objects from a 3D camera.

{{% alert title="Tip" color="tip" %}}

The following code examples assume that you have a robot configured with a [camera](/components/camera/) and a Vision Service [detector](/services/vision/detection/), [classifier](/services/vision/classification/) or [segmenter](/services/vision/segmentation/), as applicable, and that you add the required code to connect to your robot and import any required packages at the top of your code file.
Go to your robot's **Code sample** tab on the [Viam app](https://app.viam.com) for boilerplate code to connect to your robot.

{{% /alert %}}

### GetDetections

Get a list of detections from a given image using a detector.

{{< tabs >}}
{{% tab name="Python" %}}

**Parameters:**

- `image` [(RawImage)](https://python.viam.dev/autoapi/viam/media/video/index.html#viam.media.video.RawImage): The image in which to look for detections.
- `extra` (Mapping[str, Any]) (*optional*): A generic struct, containing extra options to pass to the underlying RPC call.

**Returns:**

> - [(List[Detection])](https://python.viam.dev/autoapi/viam/proto/service/vision/index.html#viam.proto.service.vision.Detectionn): A list of 2D bounding boxes, their labels, and the confidence score of the labels around the detected objects, and confidence scores of those detections.

Reviewer: ```suggestion
- [(List[Detection])](https://python.viam.dev/autoapi/viam/proto/service/vision/index.html#viam.proto.service.vision.Detection): A list of 2D bounding boxes, their labels, and the confidence score of the labels around the detected objects, and confidence scores of those detections.
```

- `count` [(int)](https://docs.python.org/3/library/functions.html#int): The number of classifications to return.
For example, if you specify `3` you will get the top three classifications with the greatest confidence scores.
- `extra` (Mapping[str, Any]) (*optional*): A generic struct, containing extra options to pass to the underlying RPC call.

**Returns:**

- [(List[Classification])](https://python.viam.dev/autoapi/viam/proto/service/vision/index.html#viam.proto.service.vision.Classification): A list of classification labels, and the confidence scores of those classifications.

For more information, see the [Python SDK Docs](https://python.viam.dev/autoapi/viam/services/vision/client/index.html#viam.services.vision.client.VisionClient.get_classifications).

```python {class="line-numbers linkable-line-numbers" data-line="11"}
# Grab camera from the robot
cam1 = Camera.from_robot(robot, "cam1")

# Get the classifier you configured on your robot
my_classifier = VisionClient.from_robot(robot, "my_classifier")

# Get an image from the camera
img = await cam1.get_image()

# Get classifications from that image

> classifications = await my_classifier.get_classifications(img)

Author: ```suggestion
classifications = await my_classifier.get_classifications(img, 2)
```

- `extra` [(map\[string\]interface{})](https://go.dev/blog/maps): Extra options to pass to the underlying RPC call.

**Returns:**

- [(classification.Classification)](https://pkg.go.dev/go.viam.com/rdk/vision/classification#Classifications): A list of classification labels, and the confidence scores of those classifications.
- [(error)](https://pkg.go.dev/builtin#error): An error, if one occurred.

For more information, see the [Go SDK Docs](https://pkg.go.dev/go.viam.com/rdk/services/vision).

```go {class="line-numbers linkable-line-numbers" data-line="22"}
// Grab the camera from the robot
cameraName := "cam1"
myCam, err := camera.FromRobot(robot, cameraName)
if err != nil {
  logger.Fatalf("cannot get camera: %v", err)
}

// Grab the classifier you configured on your robot
visService, err := vision.from_robot(robot=robot, name='my_classifier')
if err != nil {
    logger.Fatalf("Cannot get Vision Service: %v", err)
}

// Get the stream from a camera
camStream, err := myCam.Stream(context.Background())

// Get an image from the camera stream
img, release, err := camStream.Next(context.Background())
defer release()

> // Get the top 2 classifications from the image

Author: ```suggestion
// Get the 2 classifications with the highest confidence scores from the next image from the camera
```

```python {class="line-numbers linkable-line-numbers" data-line="8"}
# Grab the camera from the robot
cam1 = Camera.from_robot(robot, "cam1")

# Grab the classifier you configured on your robot
my_classifier = VisionClient.from_robot(robot, "my_classifier")

# Get the 2 classifications with the highest confidence scores from the next image from the camera
classifications = await my_classifier.get_classifications_from_camera(cam1, 2)
```

{{% /tab %}}
{{% tab name="Go" %}}

**Parameters:**

- `ctx` [(Context)](https://pkg.go.dev/context): A Context carries a deadline, a cancellation signal, and other values across API boundaries.
- `cameraName` [(string)](https://pkg.go.dev/builtin#string): The name of the camera from which to get an image to run the classifier on.
- `n` [(int)](https://pkg.go.dev/builtin#int): The number of classifications to return.
For example, if you specify `3` you will get the top three classifications with the greatest confidence scores.
- `extra` [(map\[string\]interface{})](https://go.dev/blog/maps): Extra options to pass to the underlying RPC call.

**Returns:**

> - [([]objection.Classification)](https://pkg.go.dev/go.viam.com/vision/classification#Classifications): A list of classification labels, and the confidence scores of those classifications.

Reviewer 2: ```suggestion
- [([]objection.Classification)](https://pkg.go.dev/go.viam.com/rdk/vision/classification#Classifications): A list of classification labels, and the confidence scores of those classifications.
```

- [(error)](https://pkg.go.dev/builtin#error): An error, if one occurred.

For more information, see the [Go SDK Docs](https://pkg.go.dev/go.viam.com/rdk/services/vision).

```go {class="line-numbers linkable-line-numbers" data-line="15"}
// Grab the camera from the robot
cameraName := "cam1"
myCam, err := camera.FromRobot(robot, cameraName)
if err != nil {
  logger.Fatalf("cannot get camera: %v", err)
}

// Grab the classifier you configured on your robot
visService, err := vision.from_robot(robot=robot, name='my_classifier')
if err != nil {
    logger.Fatalf("Cannot get Vision Service: %v", err)
}

> // Get the top 2 classifications from the camera output

Author: ```suggestion
// Get the 2 classifications with the highest confidence scores from the camera output
```

### GetObjectPointClouds

Get a list of 3D point cloud objects and associated metadata in the latest picture from a 3D camera (using the specified segmenter).

{{< tabs >}}
{{% tab name="Python" %}}

**Parameters:**

- `camera_name` [(string)](https://docs.python.org/3/library/stdtypes.html#text-sequence-type-str): The name of the 3D camera from which to get point cloud data.
- `extra` (Mapping[str, Any]) (*optional*): A generic struct, containing extra options to pass to the underlying RPC call.

**Returns:**

- [(List[PointCloudObject])](https://python.viam.dev/autoapi/viam/proto/common/index.html#viam.proto.common.PointCloudObject): A list of point clouds and associated metadata like the center coordinates of each point cloud.

For more information, see the [Python SDK Docs](https://python.viam.dev/autoapi/viam/services/vision/client/index.html#viam.services.vision.client.VisionClient.get_object_point_clouds).

```python {class="line-numbers linkable-line-numbers" data-line="8"}
# Grab the 3D camera from the robot
cam1 = Camera.from_robot(robot, "cam1")

# Grab the object segmenter you configured on your robot
my_segmenter = VisionClient.from_robot(robot, "my_segmenter")

> # Get the top 2 classifications from the next image from the camera

Author: ```suggestion
# Get the 2 classifications with the highest confidence scores from the next image from the camera
```

objects = await my_segmenter.get_object_point_clouds(cam1)
```

{{% /tab %}}
{{% tab name="Go" %}}

**Parameters:**

- `ctx` [(Context)](https://pkg.go.dev/context): A Context carries a deadline, a cancellation signal, and other values across API boundaries.
- `cameraName` [(string)](https://pkg.go.dev/builtin#string): The name of the 3D camera from which to get point cloud data.
- `extra` [(map\[string\]interface{})](https://go.dev/blog/maps): Extra options to pass to the underlying RPC call.

**Returns:**

> - [([]objection.Classification)](https://pkg.go.dev/go.viam.com/vision#Object): A list of point clouds and associated metadata like the center coordinates of each point cloud.

Reviewer 2: ```suggestion
- [([]objection.Classification)](https://pkg.go.dev/go.viam.com/rdk/vision#Object): A list of point clouds and associated metadata like the center coordinates of each point cloud.
```
